### PR TITLE
Allow lifetime parameters in cxx_qt::Constructors

### PR DESCRIPTION
**Pull request by LeonMatthesKDAB**
This makes it possible to have constructors take arguments by-reference, which is the default in C++.

### CODE REVIEW

**Comment by LeonMatthesKDAB** [comment]
Wondering now whether it's worth it to add this, it's such a huge pain to add, for relatively little benefit, as it's at least possible to just take most arguments by-value or use pointers as a workaround if necessary.

**Comment by ahayzen-kdab** [comment]
> Wondering now whether it's worth it to add this, it's such a huge pain to add, for relatively little benefit, as it's at least possible to just take most arguments by-value or use pointers as a workaround if necessary.

I think we need to consider the use cases where you would want to use references and there isn't a workaround. We are already quite niche though being a C++ constructor with a reference :thinking:

**Comment by LeonMatthesKDAB** [comment]
After rebasing this onto main again, the diff is actually not that bad...
The generation is still relatively complicated, so unsure if it's worth it to maintain.

I'll leave that decision up to you @Be-ing @ahayzen-kdab

File: crates/cxx-qt-gen/src/syntax/types.rs [hunk 1]
```diff
@@ -119,6 +119,85 @@ pub fn extract_qobject_ident(ty: &Type) -> Result<(Ident, Option<Mut>)> {
...
+
+fn extract_generic_argument_lifetimes(argument: &GenericArgument) -> Result<Vec<Lifetime>> {
+    match argument {
+        GenericArgument::Lifetime(lifetime) => Ok(vec![lifetime.clone()]),
+        GenericArgument::Type(ty) => extract_lifetimes(ty),
+        _ => Err(err_unsupported_type(argument)),
+    }
+}
+
+fn extract_pathsegment_lifetimes(segment: &PathSegment) -> Result<Vec<Lifetime>> {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/syntax/types.rs:134]: wonder if this stuff could be in it's own lifetimes.rs under syntax as they are sortof types helpers but very specifically lifetime helpers ?

File: crates/cxx-qt-gen/test_inputs/invokables.rs [hunk 2]
```diff
@@ -52,12 +52,14 @@ mod ffi {
...
     impl cxx_qt::Threading for MyObject {}
 
-    impl
+    impl<'a>
         cxx_qt::Constructor<
-            (i32, *mut QObject),
+            (i32, &'a mut QString),
             BaseArguments = (*mut QObject,),
-            NewArguments = (i32,),
+            NewArguments = (&'a QString,),
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/test_inputs/invokables.rs:59]: any reason this isn't `mut` ? or the other isn't const ?

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/test_inputs/invokables.rs:59]: I just wanted to show off the `default` for these two types.
E.g. the `*mut QObject` should map to a `QObject*` (usually the parent pointer). The `&'a QString` should map to a `const QString &`, which is usually what you would use in C++.

Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/test_inputs/invokables.rs:59]: But in C++ you get `QString&` for the constructor as the template type as the top is `&mut QString` ?

Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/test_inputs/invokables.rs:59]: (at least looking at invokables.h in the test outputs)

File: crates/cxx-qt-gen/src/syntax/types.rs (hunk 1 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/syntax/types.rs:134]: Hm, good point, especially for those helpers :thinking:

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/syntax/types.rs:134]: Done

File: crates/cxx-qt-gen/test_inputs/invokables.rs (hunk 2 above)
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/test_inputs/invokables.rs:59]: Aaah, that `mut` you mean.
Good point.
Fixed.

**Review by ahayzen-kdab — APPROVED** [review]
LGTM, lets go with this for now, in the end not too crazy :-) We can always strip it back out before the release if we find something that doesn't work.

**Pull request merged**